pallets/dynamic-evm-base-fee/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+
+pub use self::pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+// TODO: remove this after proper benchmarking!
+pub trait WeightInfo {
+    fn base_fee_per_gas_adjustment() -> Weight;
+    fn set_base_fee_per_gas() -> Weight;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+    use frame_support::pallet_prelude::*;
+    use frame_system::pallet_prelude::*;
+
+    use super::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Overarching event type
+        type RuntimeEvent: From<Event> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        /// Default base fee per gas value. Used in genesis if no other value specified explicitly.
+        type DefaultBaseFeePerGas: Get<U256>;
+        /// Minimum value 'base fee per gas' can be adjusted to. This is a defensive measure to prevent the fee from being too low.
+        type MinBaseFeePerGas: Get<U256>;
+        /// Maximum value 'base fee per gas' can be adjusted to. This is a defensive measure to prevent the fee from being too high.
+        type MaxBaseFeePerGas: Get<U256>;
+        /// Getter for the fee adjustment factor used in 'base fee per gas' formula. This is expected to change in-between the blocks (doesn't have to though).
+        type AdjustmentFactor: Get<FixedU128>;
+        /// The so-called `weight_factor` in the 'base fee per gas' formula.
+        type WeightFactor: Get<u128>;
+        /// Ratio limit on how much the 'base fee per gas' can change in-between two blocks.
+        /// It's expressed as percentage, and used to calculate the delta between the old and new value.
+        /// E.g. if the current 'base fee per gas' is 100, and the limit is 10%, then the new base fee per gas can be between 90 and 110.
+        type StepLimitRatio: Get<Perquintill>;
+        /// Weight information for extrinsics & functions of this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::type_value]
+    pub fn DefaultBaseFeePerGas<T: Config>() -> U256 {
+        T::DefaultBaseFeePerGas::get()
+    }
+
+    #[pallet::storage]
+    pub type BaseFeePerGas<T> = StorageValue<_, U256, ValueQuery, DefaultBaseFeePerGas<T>>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event {
+        /// New `base fee per gas` value has been force-set.
+        NewBaseFeePerGas { fee: U256 },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Specified value is outside of the allowed range.
+        ValueOutOfBounds,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_: T::BlockNumber) -> Weight {
+            T::WeightInfo::base_fee_per_gas_adjustment()
+        }
+
+        fn on_finalize(_n: <T as frame_system::Config>::BlockNumber) {
```

Comment:
After updating the bfpg `on_finialize` should also emit `NewBaseFeePerGas` event with updated value for visibility.

Reply:
No, we don't want to bloat the state. This will be done on every single block.
Event is added just for extrinsic on purpose.

A lot of things get updated in the block hooks but we don't emit events for those.